### PR TITLE
[SPARK-28700][SQL] Use DECIMAL type for `sec` in `make_timestamp()`

### PR DESCRIPTION
**Pull request by MaxGekk**
## What changes were proposed in this pull request?

Changed type of `sec` argument in the `make_timestamp()` function from `DOUBLE` to `DECIMAL(8, 6)`. The scale is set to 6 to cover microsecond fractions, and the precision is 2 digits for seconds + 6 digits for microsecond fraction. New type prevents losing precision in some cases, for example:

Before:
```sql
spark-sql> select make_timestamp(2019, 8, 12, 0, 0, 58.000001);
2019-08-12 00:00:58
```

After:
```sql
spark-sql> select make_timestamp(2019, 8, 12, 0, 0, 58.000001);
2019-08-12 00:00:58.000001
```

Also switching to `DECIMAL` fixes rounding `sec` towards "nearest neighbor" unless both neighbors are equidistant, in which case round up. For example:

Before:
```sql
spark-sql> select make_timestamp(2019, 8, 12, 0, 0, 0.1234567);
2019-08-12 00:00:00.123456
```

After:
```sql
spark-sql> select make_timestamp(2019, 8, 12, 0, 0, 0.1234567);
2019-08-12 00:00:00.123457
```

## How was this patch tested?

This was tested by `DateExpressionsSuite` and `pgSQL/timestamp.sql`.

### CODE REVIEW

**Comment by MaxGekk** [comment]
@srowen @cloud-fan @HyukjinKwon @dongjoon-hyun Please, review the PR.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala [hunk 1]
```diff
@@ -388,10 +388,12 @@ final class Decimal extends Ordered[Decimal] with Serializable {
       Decimal(toBigDecimal - that.toBigDecimal)
     }
   }
+  def sub(that: Decimal): Decimal = this - that
 
   // TypeCoercion will take care of the precision, scale of result
   def * (that: Decimal): Decimal =
     Decimal(toJavaBigDecimal.multiply(that.toJavaBigDecimal, MATH_CONTEXT))
+  def mul(that: Decimal): Decimal = this * that
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: BTW I think you can call the existing operator with `.$times()` in Java code

**Review by srowen — COMMENTED** [review]
I agree with the change and trust your judgment. It should be mentioned in the migration guide, maybe? as it is a tiny behavior change.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala (hunk 1 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/types/Decimal.scala]: @srowen Thank you for this. I replaced the methods by `$minus` and `$times` as you suggested.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. `make_timestamp` is a new function added at 3.0.0. (So, we don't need to add to the migration guide)

Thank you, @MaxGekk and @srowen .

Merged to master.

**Comment by HyukjinKwon** [comment]
+1 too

**Pull request closed without merging**